### PR TITLE
Feature/Store Kit Configuration

### PR DESCRIPTION
**Pull request by jcolicchio**
Addresses issue #944 

This PR adds `storeKitConfiguration` as an optional property of the Run Scheme and TargetScheme

In the yaml we can pass the path to the `.storekit` as a string or a dictionary with `location` and optional `forWorkspace`.

`forWorkspace` is required in order to support usage with `xcworkspace`s

### CODE REVIEW

File: Sources/XcodeGenKit/SchemeGenerator.swift
Comment (jcolicchio) [inline Sources/XcodeGenKit/SchemeGenerator.swift:191]: This change was required as part of the upgrade to `XcodeProj 7.15.0`

**Comment by jcolicchio** [comment]
@yonaskolb Passing tests and ready for review!

File: CHANGELOG.md
Comment (gemmakbarlow) [inline CHANGELOG.md:7]:
```suggestion
- Add `Scheme.StoreKitConfiguration` to allow specifying StoreKit Configuration in Scheme and TargetScheme, supporting either xcodeproj or xcworkspace. [#964](https://github.com/yonaskolb/XcodeGen/pull/964) @jcolicchio
```

File: Docs/ProjectSpec.md
Comment (gemmakbarlow) [inline Docs/ProjectSpec.md:840]: Nice.

File: Sources/ProjectSpec/Scheme.swift
Comment (gemmakbarlow) [inline Sources/ProjectSpec/Scheme.swift:416]:
```suggestion
        storeKitConfiguration = maybeStoreKitLocation ?? maybeStoreKitLocation.flatMap { Scheme.StoreKitConfiguration(location: $0) }
```

File: Sources/ProjectSpec/TargetScheme.swift
Comment (gemmakbarlow) [inline Sources/ProjectSpec/TargetScheme.swift:75]:
```suggestion
        storeKitConfiguration = maybeStoreKitLocation ?? maybeStoreKitLocation.flatMap { Scheme.StoreKitConfiguration(location: $0) }
```

**Review by gemmakbarlow — APPROVED** [review]
Looks great ✅ 

Some very minor suggestions below, but none that should block this from being merged. See what you think !

File: CHANGELOG.md
Comment (jcolicchio) [reply CHANGELOG.md:7]: I can make this change, but I think it's an important distinction that a scheme can support one or the other, but not both simultaneously

Comment (jcolicchio) [reply CHANGELOG.md:7]: Oops I'm dumb, I read the diff backwards 😨

File: Sources/ProjectSpec/Scheme.swift
Comment (jcolicchio) [reply Sources/ProjectSpec/Scheme.swift:416]: Though it looks redundant, `jsonDictionary.json(atKeyPath: "storeKitConfiguration")` on L416 is actually attempting to decode "storeKitConfiguration" as a `Scheme.StoreKitConfiguration`, whereas on L415 it's attempting to decode it as a `String`. I've pushed a commit cleaning this up and moving it into `Scheme.StoreKitConfiguration`'s init, let me know what you think!

Comment (gemmakbarlow) [reply Sources/ProjectSpec/Scheme.swift]: Whoops, got it. Nice, this adjustment looks solid 👍🏻

File: Sources/ProjectSpec/Scheme.swift
Comment (gemmakbarlow) [inline Sources/ProjectSpec/Scheme.swift:76]: ❤️

**Comment by jcolicchio** [comment]
@yonaskolb this is ready for review and needs one more reviewer if you'd like to do the honors. Moving forward, is there a process for PRs like this? Should I just tag you when something needs attention?

File: Sources/ProjectSpec/Scheme.swift
Comment (yonaskolb) [inline Sources/ProjectSpec/Scheme.swift:84]: Are we sure the prefixes will never change or are other things in different scenarios? Maybe we should rename this to prefix, give it a default of "../../" and add info in the docs that it must be changed for workspaces. Thoughts?

Comment (yonaskolb) [reply Sources/ProjectSpec/Scheme.swift:84]: Personally I would have left has as a single `storekitConfiguration` string that required putting the prefix in and documented the usage of it. I'm hesitant to put too many smarts in XcodeGen for small side features, especially anything in Schemes as they change so much. But you've done the hard work here to make it more friendly already. If it's left in, it would be good to have some tests for the prefix changing with the boolean

Comment (yonaskolb) [reply Sources/ProjectSpec/Scheme.swift:84]: As XcodeGen gets older I have a greater appreciation for small features that don't need to updated when Xcode does 😄

Comment (jcolicchio) [reply Sources/ProjectSpec/Scheme.swift:84]: I absolutely see your point here, and I don't want to ship something that'll break with no recourse the next time an Xcode internal detail changes. How about this: instead of `forWorkspace`, I'll add `pathPrefix` with a default value of `../../`? `SimulateLocation` could also stand to benefit from this addition as it's currently broken for non-workspace projects...

Comment (jcolicchio) [reply Sources/ProjectSpec/Scheme.swift:84]: This way, the path can still be relative to the project root, same as other sources, but for `.xcworkspace` projects or when Xcode breaks, we can tweak as needed by setting `pathPrefix`? I'd just like for XcodeGen users not to have to worry about this weird wonky path detail until they actually need to

Comment (jcolicchio) [reply Sources/ProjectSpec/Scheme.swift]: @yonaskolb Ready for another round of review, does this look better?

**Comment by jcolicchio** [comment]
@yonaskolb just merged master, assuming this still passes tests it could use some attention :0

**Comment by jcolicchio** [comment]
@yonaskolb It seems like you're more active on GitHub now -- this PR has been open for a few months, can I get a quick re-review? Or is there someone else I can ping to get it reviewed and merged?

**Comment by yonaskolb** [comment]
Hi @jcolicchio, apologies for the delay here. I've had another look through this, and have some follow up thoughts. In an effort to both simplify the `storekitConfiguration` and to make it easier to change the path for all `storekitConfiguration`s in a spec, what if the prefix was in the top level options, and `storekitConfiguration` was just a string. Then it could also be used for `SimulateLocation` as you say. Not sure of naming, `workspacePrefix`? Thoughts?
I promise this will get actioned immediately, and won't be left for ages again :)

**Comment by jcolicchio** [comment]
> Hi @jcolicchio, apologies for the delay here. I've had another look through this, and have some follow up thoughts. In an effort to both simplify the `storekitConfiguration` and to make it easier to change the path for all `storekitConfiguration`s in a spec, what if the prefix was in the top level options, and `storekitConfiguration` was just a string. Then it could also be used for `SimulateLocation` as you say. Not sure of naming, `workspacePrefix`? Thoughts?
> I promise this will get actioned immediately, and won't be left for ages again :)

I think that's a great idea, and dovetails nicely into a similar fix affecting simulated locations. As for naming, I'll give it some thought. Ultimately, it's a prefix for relative paths within schemes, so maybe `schemePathPrefix` might be good?

**Comment by jcolicchio** [comment]
@yonaskolb I've changed `storeKitConfiguration` to a string, added an option for path prefix (name negotiable and easily find/replaceable), updated project spec and tests, we're good to go if the name looks ok 👍

**Review by yonaskolb — APPROVED** [review]
Fantastic, great name and great work! 👏

File: Sources/ProjectSpec/FileType.swift
Comment (liamnichols) [inline Sources/ProjectSpec/FileType.swift:113]: 👋 @jcolicchio, please can you help me understand this addition quickly? 

When using a **.storekit** file in a scheme I think maybe makes sense that this wouldn't be included as a resource, however when updating to 2.19.0 today, my tested started failing because `SKTestSession(configurationFileNamed: "StoreKitConfiguration")` couldn't find my configuration file anymore. 

Could this have maybe been an oversight because to use the **.storekit** files with `SKTestSession` then they need to be bundled with the test target? 

Or is it that Xcode doesn't treat **.storekit** files as resources by default? I haven't actually tested adding one to Xcode manually to see how it behaves out of the box so maybe the prior behaviour was a side effect that I had been relying on. 

I can solve it in my spec file by overriding the `buildRule` in `sources`, but I'm wondering if maybe we want to refer this or thing actually 🤔

Comment (jcolicchio) [reply Sources/ProjectSpec/FileType.swift:113]: Hey @liamnichols , sorry to hear my recent change has broken your setup! I'll look into how Xcode typically deals with *.storekit* files, it may indeed make sense for them to be copied by default. My use case involved referencing a storekit file in the scheme, which doesn't appear to require the *.storekit* file be copied. Apologies!

Comment (liamnichols) [reply Sources/ProjectSpec/FileType.swift:113]: No stress, thanks for answering my question so quickly 🙌 

TBH I was also only just finding out that **.storekit** files could be used in two separate ways 😅 I only noticed that our project was doing this because our tests started failing when I upgraded XcodeGen 😬

Comment (jcolicchio) [reply Sources/ProjectSpec/FileType.swift:113]: Hah all good, I guess we both learned something today :P

**Pull request merged**
